### PR TITLE
BUGFIX: Fix `contentgraphintegrity:runviolationdetection`

### PR DESCRIPTION
**Pull request by mhsdesign**
The wiring for `flow contentgraphintegrity:runviolationdetection` was wrong and needed to be redone.

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by kitsunet** [comment]
I would be fine merging this although silently different behavior due to different environments has bitten us a couple of times before, so IDK, maybe there is a way to make this query backwards compatible?

**Comment by mhsdesign** [comment]
Hmm im not that deep into json functions for maria db. I thought for a second that performance will not matter as this is just run for the behat tests but as i found out there is a cli command that is supposed to be run on large sets of data so better keep it high performing.

... we could add a switch to use php filtering if the json functions are not available but hard to test.

Id say this is just a hack and we can even drop it before final release, when we are sure what we want to support: https://github.com/neos/neos-development-collection/issues/4337

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/ProjectionIntegrityViolationDetector.php [hunk 1]
```diff
@@ -182,20 +182,25 @@ public function subtreeTagsAreInherited(): Result
...
+        WHERE
+          EXISTS (
+            SELECT t.tag FROM JSON_TABLE(JSON_KEYS(ph.subtreetags), \'$[*]\' COLUMNS(tag VARCHAR(30) PATH \'$\')) t WHERE NOT JSON_EXISTS(h.subtreetags, CONCAT(\'$.\', t.tag))
+          )';
+
+        try {
+            $hierarchyRelationsWithMissingSubtreeTags = $this->client->getConnection()->executeQuery($query)->fetchAllAssociative();
+        } catch (\Doctrine\DBAL\Exception\SyntaxErrorException $syntaxErrorException) {
+            // Hierarchy relation subtree tags could not be validated as the database doesn't support "JSON_TABLE" feature.
+            return $result;
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/ProjectionIntegrityViolationDetector.php]:
```suggestion
            // Hierarchy relation subtree tags could not be validated as the database doesn't support "JSON_TABLE" feature.
            // TODO drop this check once the db requirements are clear: https://github.com/neos/neos-development-collection/issues/4337
            return $result;
```

File: Neos.ContentRepositoryRegistry/Classes/Command/ContentGraphIntegrityCommandController.php [hunk 2]
```diff
@@ -11,29 +11,35 @@
...
-    public function __construct(ProjectionIntegrityViolationDetectionRunner $detectionRunner)
+    public function runViolationDetectionCommand(string $contentRepository = 'default', string $outputMode = null): void
     {
-        $this->detectionRunner = $detectionRunner;
-        parent::__construct();
-    }
+        $detectionRunner = $this->contentRepositoryRegistry->buildService(
+            ContentRepositoryId::fromString($contentRepository),
+            new DoctrineDbalProjectionIntegrityViolationDetectionRunnerFactory($this->dbalClient)
+        );
```
Comment (bwaidelich) [inline Neos.ContentRepositoryRegistry/Classes/Command/ContentGraphIntegrityCommandController.php:38]: (How) is this change related to the bug?

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/ProjectionIntegrityViolationDetector.php [hunk 3]
```diff
@@ -182,20 +182,26 @@ public function subtreeTagsAreInherited(): Result
...
+        FROM
+          ' . $this->tableNamePrefix . '_hierarchyrelation h
+          INNER JOIN ' . $this->tableNamePrefix . '_hierarchyrelation ph
+            ON ph.childnodeanchor = h.parentnodeanchor
+            AND ph.contentstreamid = h.contentstreamid
+            AND ph.dimensionspacepointhash = h.dimensionspacepointhash
+        WHERE
+          EXISTS (
+            SELECT t.tag FROM JSON_TABLE(JSON_KEYS(ph.subtreetags), \'$[*]\' COLUMNS(tag VARCHAR(30) PATH \'$\')) t WHERE NOT JSON_EXISTS(h.subtreetags, CONCAT(\'$.\', t.tag))
+          )';
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/ProjectionIntegrityViolationDetector.php]: nitpick: It makes sense to put this into a variable, but the indentation is wrong, I would suggest to use Nowdoc like
$query = <<<'SQL'
...
SQL;

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/ProjectionIntegrityViolationDetector.php [hunk 4]
```diff
@@ -182,20 +182,26 @@ public function subtreeTagsAreInherited(): Result
...
+
+        try {
+            $hierarchyRelationsWithMissingSubtreeTags = $this->client->getConnection()->executeQuery($query)->fetchAllAssociative();
+        } catch (\Doctrine\DBAL\Exception\SyntaxErrorException $syntaxErrorException) {
+            // Hierarchy relation subtree tags could not be validated as the database doesn't support "JSON_TABLE" feature.
+            // TODO drop this check once the db requirements are clear: https://github.com/neos/neos-development-collection/issues/4337
+            return $result;
+        }
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/ProjectionIntegrityViolationDetector.php]: This will forever render this test useless if we happen to have a typo in the query above.
We should instead:
a) rewrite the query such that it works on older versions (maybe with a slight performance penalty)
b) ...or require MariaDB 10.6+ (which is required for [JSON_TABLE](https://mariadb.com/kb/en/json_table/) to work) – this version was released [3 years ago](https://mariadb.com/kb/en/mariadb-1060-release-notes/)

**Review by bwaidelich — DISMISSED** [review]

**Comment by mhsdesign** [comment]
Okay this was rebased and refactored, this change now only contains minimal adjustments to rewire the command as otherwise it will crash (look at the code :D)

**Comment by mhsdesign** [comment]
Merging this as this is now a simple adjustment as trival as phpstan lint errors (actually would be cool if phpstan could now what is autowireable and what not :D)
